pkg/cmd/pulumi/package_gen_sdk.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2016-2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
+
+	javagen "github.com/pulumi/pulumi-java/pkg/codegen/java"
+
+	"github.com/pulumi/pulumi/pkg/v3/codegen/dotnet"
+	gogen "github.com/pulumi/pulumi/pkg/v3/codegen/go"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/nodejs"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/python"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+)
+
+func newGenSdkCommand() *cobra.Command {
+	var language string
+	var out string
+	cmd := &cobra.Command{
+		Use:   "gen-sdk <schema_source>",
+		Args:  cobra.ExactArgs(1),
+		Short: "Generate SDK(s) from a package or schema",
+		Long: `Generate SDK(s) from a package or schema.
+
+<schema_source> can be a package name, the path to a plugin binary, or the path to a schema file.`,
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+			source := args[0]
+
+			pkg, err := schemaFromSchemaSource(source)
+			if err != nil {
+				return err
+			}
+			if language == "all" {
+				for _, lang := range []string{"dotnet", "go", "java", "nodejs", "python"} {
+					err := genSDK(lang, out, pkg)
+					if err != nil {
+						return err
+					}
+				}
+				return nil
+			}
+			return genSDK(language, out, pkg)
+		}),
+	}
+	cmd.Flags().StringVarP(&language, "language", "", "all",
+		"The SDK language to generate: [nodejs|python|go|dotnet|java|all]")
+	cmd.Flags().StringVarP(&out, "out", "o", "./sdk",
+		"The directory to write the SDK to")
+	return cmd
+}
+
+func genSDK(language, out string, pkg *schema.Package) error {
+	var f func(string, *schema.Package, map[string][]byte) (map[string][]byte, error)
+	switch language {
```

Comment:
I wonder if this functionality should live in the language plugins. Then we wouldn't need to hard code this `switch`, we'd just load up the requested language plugin and call it over gRPC to do the SDK codegen, similar in spirit to #9294. The benefit being that if we add more languages in the future, we wouldn't necessarily have to touch this code again.

Not sure what that means for `all`, though, other than enumerating all available language plugins and invoking the appropriate gRPC method for each one.

Though, I can also understand possible pushback to this suggestion: that the codegen functionality doesn't really belong in the language plugins because their purpose (right now) is only about running Pulumi programs in the given language -- not about codegen. If others share this concern, we could consider another kind of plugin, maybe "codegen" (e.g. `pulumi-codegen-nodejs`, `pulumi-codegen-python`, etc.) that would be exclusively meant to handle codegen tasks (perhaps both sdkgen and programgen).

Of course, we could always hard code in a `switch` for now for expediency, and move to language plugin in the future.

Reply:
Codegen can and should be pluginafied, but that is out of scope of this PR. Were hardcoding for now, but that is not the desired end goal.

Reply:
>I wonder if this functionality should live in the language plugins. 

It should! We have discussed this multiple times, we will do this but it's probably tricky so keeps being put off.